python/pyspark/sql/types.py

```diff
@@ -1468,6 +1475,36 @@ def __eq__(self, other: Any) -> bool:
         return type(self) == type(other)
 
 
+class VariantVal:
```

Comment:
some types including `YearMonthIntervalType/CalendarIntervalType` do not have the corresponding value classes, since only the builtin python classes are accepted before.

Should we support value class in this case? If so, should we support those types in this way? @HyukjinKwon

Reply:
`Row` is here too though. But I don't mind either way

Reply:
I'm not sure what this comment means?

Reply:
I think he was suggesting to put `VariantVal` into somewhere else e.g., making a separate file `vairent.py`, and put all related things into that.

Reply:
BTW, if we want this to be documented, we should list it under `python/docs/source/reference/pyspark.sql`.

Reply:
Added to the documentation.

I think for now, keeping the lightweight `VariantVal` here, and keeping all the binary encoding details in `variant_utils.py` seems fine. We can refactor later if needed.